xdsl/transforms/csl_stencil_bufferize.py

```diff
@@ -371,13 +379,77 @@ def match_and_rewrite(self, op: arith.Constant, rewriter: PatternRewriter, /):
         )
 
 
+@dataclass(frozen=True)
+class ReselectLinalgOutsFromInputs(RewritePattern):
+    """
+    Reselects the DPS `outs` of a linalg op if it is one of its inputs.
+      * select `writable` tensor input with no later use, or else
+      * select linalg op input with no later use
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: linalg.NamedOpBase | linalg.Generic, rewriter: PatternRewriter, /
+    ):
+        # only apply rewrite when re-selecting `outs` from `ins`
+        if (
+            op.outputs[0] not in op.inputs
+            or self.is_writable(op.outputs[0])
+            or len(op.outputs) != 1
+        ):
+            return
+
+        # the new `outs` to re-select
+        out: SSAValue | None = None
+
+        for arg in op.inputs:
+            # reselect outs that has no later use to avoid read-after-write conflicts
+            if len(arg.uses) == 1:
+                # check for a `writable` input with no later uses and break immediately
+                if self.is_writable(arg):
+                    out = arg
+                    break
+
+                # check for a linalg op input with no later uses and keep looking
+                if isinstance(arg, OpResult) and isinstance(
+                    arg.op, linalg.NamedOpBase | linalg.Generic
+                ):
+                    out = arg
```

Comment:
If input is not writable, why do we still make it the output? Since we have asserted that at this point the output must be one of the inputs and not writable. Would this not needlessly rebuild the op?

Reply:
That makes sense. Although I would've expected `mlir` to figure that out when it does the bufferisation

Reply:
It doesn't really do much analysis, and generally has two options it considers: re-use the buffer given as outs, or create a copy. Chaining ops, such as providing the result of a prev op as outs (not: the outs of a prev op) helps with the first option. But if outs is a tensor from an arith.const, it creates a buffer copy. Try it on https://godbolt.org/z/YjrTG54oP